In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [72]:
import pandas as pd
import numpy as np
import os
import re
from bs4 import BeautifulSoup
import requests
from operator import itemgetter 

In [3]:
data=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
data

'<!DOCTYPE html>\n<html class="client-nojs" lang="en" dir="ltr">\n<head>\n<meta charset="UTF-8"/>\n<title>List of postal codes of Canada: M - Wikipedia</title>\n<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"63cc7353-832f-43ec-a9dc-49dca7add404","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":979555370,"wgRevisionId":979555370,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ontar

In [4]:
data

'<!DOCTYPE html>\n<html class="client-nojs" lang="en" dir="ltr">\n<head>\n<meta charset="UTF-8"/>\n<title>List of postal codes of Canada: M - Wikipedia</title>\n<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"63cc7353-832f-43ec-a9dc-49dca7add404","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":979555370,"wgRevisionId":979555370,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ontar

In [5]:
data_soup=BeautifulSoup(data, "lxml")
# soem other parsers=specific parser ("lxml","lxml-xml", "html.parser", or "html5lib") or it may be the type of markup to be
#used ("html", "html5", "xml"). It's recommended that you name a specific parser, so that Beautiful Soup gives you the same
#results across platforms and virtual environments.
print(data_soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"63cc7353-832f-43ec-a9dc-49dca7add404","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":979555370,"wgRevisionId":979555370,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communicati

In [6]:
data_soup

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"63cc7353-832f-43ec-a9dc-49dca7add404","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":979555370,"wgRevisionId":979555370,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ontario","P

In [7]:
print(data_soup.title)

<title>List of postal codes of Canada: M - Wikipedia</title>


In [8]:
print(data_soup.title.text)

List of postal codes of Canada: M - Wikipedia


In [9]:
for l in data_soup.find_all("a"):
    print("Inner Text: {}".format(l.text))
    print("Title: {}".format(l.get("title")))
    print("href: {}".format(l.get("href")))

Inner Text: 
Title: None
href: None
Inner Text: Jump to navigation
Title: None
href: #mw-head
Inner Text: Jump to search
Title: None
href: #searchInput
Inner Text: postal codes in Canada
Title: Postal codes in Canada
href: /wiki/Postal_codes_in_Canada
Inner Text: Toronto
Title: Toronto
href: /wiki/Toronto
Inner Text: Ontario
Title: Ontario
href: /wiki/Ontario
Inner Text: Canada Post
Title: Canada Post
href: /wiki/Canada_Post
Inner Text: [1]
Title: None
href: #cite_note-1
Inner Text: applications
Title: Mobile app
href: /wiki/Mobile_app
Inner Text: smartphones
Title: Smartphones
href: /wiki/Smartphones
Inner Text: iPhone
Title: IPhone
href: /wiki/IPhone
Inner Text: BlackBerry
Title: BlackBerry
href: /wiki/BlackBerry
Inner Text: [2]
Title: None
href: #cite_note-2
Inner Text: CD-ROMs
Title: CD-ROM
href: /wiki/CD-ROM
Inner Text: Toronto
Title: Toronto
href: /wiki/Toronto
Inner Text: FSAs
Title: Postal codes in Canada
href: /wiki/Postal_codes_in_Canada#Forward_sortation_areas
Inner Text: ed

### Test code
- #### Go down to after test code to see the actual code written in a  order

In [10]:
table_req = data_soup.find("div", attrs={"class": "mw-parser-output"})
table_req

<div class="mw-parser-output"><div class="shortdescription nomobile noexcerpt noprint searchaux" style="display:none">Wikipedia list article</div>
<p>This is a list of <a href="/wiki/Postal_codes_in_Canada" title="Postal codes in Canada">postal codes in Canada</a> where the first letter is M. Postal codes beginning with M are located within the city of <a href="/wiki/Toronto" title="Toronto">Toronto</a> in the province of <a href="/wiki/Ontario" title="Ontario">Ontario</a>. Only the first three characters are listed, corresponding to the Forward Sortation Area.
</p><p><a href="/wiki/Canada_Post" title="Canada Post">Canada Post</a> provides a free postal code look-up tool on its website,<sup class="reference" id="cite_ref-1"><a href="#cite_note-1">[1]</a></sup> via its <a href="/wiki/Mobile_app" title="Mobile app">applications</a> for such <a class="mw-redirect" href="/wiki/Smartphones" title="Smartphones">smartphones</a> as the <a href="/wiki/IPhone" title="IPhone">iPhone</a> and <a hr

In [11]:
table_data = table_req.tbody.find_all("th")
table_data

[<th>Postal Code
 </th>,
 <th>Borough
 </th>,
 <th>Neighbourhood
 </th>]

In [12]:
table_data[0]

<th>Postal Code
</th>

In [14]:
table_data[1]

<th>Borough
</th>

In [17]:
table_req

<div class="mw-parser-output"><div class="shortdescription nomobile noexcerpt noprint searchaux" style="display:none">Wikipedia list article</div>
<p>This is a list of <a href="/wiki/Postal_codes_in_Canada" title="Postal codes in Canada">postal codes in Canada</a> where the first letter is M. Postal codes beginning with M are located within the city of <a href="/wiki/Toronto" title="Toronto">Toronto</a> in the province of <a href="/wiki/Ontario" title="Ontario">Ontario</a>. Only the first three characters are listed, corresponding to the Forward Sortation Area.
</p><p><a href="/wiki/Canada_Post" title="Canada Post">Canada Post</a> provides a free postal code look-up tool on its website,<sup class="reference" id="cite_ref-1"><a href="#cite_note-1">[1]</a></sup> via its <a href="/wiki/Mobile_app" title="Mobile app">applications</a> for such <a class="mw-redirect" href="/wiki/Smartphones" title="Smartphones">smartphones</a> as the <a href="/wiki/IPhone" title="IPhone">iPhone</a> and <a hr

In [18]:
#table_req = data_soup.find("div", attrs={"class": "mw-parser-output"})
table_req = data_soup.find("div", attrs={"class": "mw-parser-output"})
#table_req
table_data = table_req.tbody.find_all("tr")  # contains many rows

In [19]:
table_data

[<tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighbourhood
 </th></tr>,
 <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>,
 <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>,
 <tr>
 <td>M5A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Regent Park, Harbourfront
 </td></tr>,
 <tr>
 <td>M6A
 </td>
 <td>North York
 </td>
 <td>Lawrence Manor, Lawrence Heights
 </td></tr>,
 <tr>
 <td>M7A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Queen's Park, Ontario Provincial Government
 </td></tr>,
 <tr>
 <td>M8A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M9A
 </td>
 <td>Etobicoke
 </td>
 <td>Islington Avenue, Humber Valley Village
 </td></tr>,
 <tr>
 <td>M1B
 </td>
 <td>Scarborough
 </td>
 <td>Malvern, Rouge
 </td></tr>,
 <tr>
 <td>M2B
 </td>
 <td>Not assigned
 </td>

In [20]:
table_data[0]

<tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>

In [21]:
# Get all the headings of Lists
headings = []
for td in table_data[0].find_all("th"):
    # remove any newlines and extra spaces from left and right
    headings.append(td.text.replace('\n', ' ').strip())
    #headings.append(td.b.text.replace('\n', ' ').strip())
print(headings)

['Postal Code', 'Borough', 'Neighbourhood']


In [22]:
table_data[1]

<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>

In [23]:
table_data

[<tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighbourhood
 </th></tr>,
 <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>,
 <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>,
 <tr>
 <td>M5A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Regent Park, Harbourfront
 </td></tr>,
 <tr>
 <td>M6A
 </td>
 <td>North York
 </td>
 <td>Lawrence Manor, Lawrence Heights
 </td></tr>,
 <tr>
 <td>M7A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Queen's Park, Ontario Provincial Government
 </td></tr>,
 <tr>
 <td>M8A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M9A
 </td>
 <td>Etobicoke
 </td>
 <td>Islington Avenue, Humber Valley Village
 </td></tr>,
 <tr>
 <td>M1B
 </td>
 <td>Scarborough
 </td>
 <td>Malvern, Rouge
 </td></tr>,
 <tr>
 <td>M2B
 </td>
 <td>Not assigned
 </td>

In [24]:
for tr in table_data[1]:
    print(tr)



<td>M1A
</td>


<td>Not assigned
</td>


<td>Not assigned
</td>


In [25]:
data1 = {}
data_rows = []
for tr in table_data: # find all tr's from table's tbody
    t_row = {}
    # Each table row is stored in the form of a dictionary
       # find all td's in tr and zip it with headings
    for td, th in zip(tr.find_all("td"), headings): 
        t_row[th] = td.text.replace('\n', '').strip()
    data_rows.append(t_row)
print(data_rows)

[{}, {'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}, {'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}, {'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}, {'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}, {'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}, {'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}, {'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}, {'Postal Code': 'M2B', 'Borough': 'Not assigned', 'Neighbourhoo

In [26]:
headings

['Postal Code', 'Borough', 'Neighbourhood']

In [27]:
data1[td] = data_rows
print(data1)

{<td>Not assigned
</td>: [{}, {'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}, {'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}, {'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}, {'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}, {'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}, {'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}, {'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}, {'Postal Code': 'M2B', 'Borough': 'Not

In [28]:
data1 = {}
data_rows = []
for tr in table_data: # find all tr's from table's tbody
    t_row = {}
    # Each table row is stored in the form of a dictionary
       # find all td's in tr and zip it with headings
    for td, th in zip(tr.find_all("td"), headings): 
        t_row[th] = td.text.replace('\n', '').strip()
    data_rows.append(t_row)
    # Put the data for the table with his heading.
data1[td] = data_rows
print(data1)

{<td>Not assigned
</td>: [{}, {'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}, {'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}, {'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}, {'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}, {'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}, {'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}, {'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}, {'Postal Code': 'M2B', 'Borough': 'Not

In [29]:
type(data_rows)
type(headings)
type(td)

list

list

bs4.element.Tag

In [30]:
len(data_rows)
len(headings)

181

3

In [31]:
data_rows[1]

{'Postal Code': 'M1A',
 'Borough': 'Not assigned',
 'Neighbourhood': 'Not assigned'}

In [32]:
type(data_rows[1])

dict

In [33]:
com=[]
for x in data_rows[1]["Postal Code"]:
    com.append(x)
    post_rows=' '.join(com)
print(com)
print(post_rows)

['M', '1', 'A']
M 1 A


In [62]:
com1=[]
def firstextract(n):
    for x in data_rows[n]["Postal Code"]:
        com1.append(x)
        post_rows1=' '.join(com1)
    return com1
    return post_rows1

In [35]:
firstextract(5)

['M', '5', 'A']

In [90]:
extract1=[d['Postal Code'] for d in data_rows if 'Postal Code' in d]
extract2=[d['Borough'] for d in data_rows if 'Borough' in d]
extract3=[d['Neighbourhood'] for d in data_rows if 'Neighbourhood' in d]

In [92]:
print(extract1)
print(extract2)
print(extract3)

['M1A', 'M2A', 'M3A', 'M4A', 'M5A', 'M6A', 'M7A', 'M8A', 'M9A', 'M1B', 'M2B', 'M3B', 'M4B', 'M5B', 'M6B', 'M7B', 'M8B', 'M9B', 'M1C', 'M2C', 'M3C', 'M4C', 'M5C', 'M6C', 'M7C', 'M8C', 'M9C', 'M1E', 'M2E', 'M3E', 'M4E', 'M5E', 'M6E', 'M7E', 'M8E', 'M9E', 'M1G', 'M2G', 'M3G', 'M4G', 'M5G', 'M6G', 'M7G', 'M8G', 'M9G', 'M1H', 'M2H', 'M3H', 'M4H', 'M5H', 'M6H', 'M7H', 'M8H', 'M9H', 'M1J', 'M2J', 'M3J', 'M4J', 'M5J', 'M6J', 'M7J', 'M8J', 'M9J', 'M1K', 'M2K', 'M3K', 'M4K', 'M5K', 'M6K', 'M7K', 'M8K', 'M9K', 'M1L', 'M2L', 'M3L', 'M4L', 'M5L', 'M6L', 'M7L', 'M8L', 'M9L', 'M1M', 'M2M', 'M3M', 'M4M', 'M5M', 'M6M', 'M7M', 'M8M', 'M9M', 'M1N', 'M2N', 'M3N', 'M4N', 'M5N', 'M6N', 'M7N', 'M8N', 'M9N', 'M1P', 'M2P', 'M3P', 'M4P', 'M5P', 'M6P', 'M7P', 'M8P', 'M9P', 'M1R', 'M2R', 'M3R', 'M4R', 'M5R', 'M6R', 'M7R', 'M8R', 'M9R', 'M1S', 'M2S', 'M3S', 'M4S', 'M5S', 'M6S', 'M7S', 'M8S', 'M9S', 'M1T', 'M2T', 'M3T', 'M4T', 'M5T', 'M6T', 'M7T', 'M8T', 'M9T', 'M1V', 'M2V', 'M3V', 'M4V', 'M5V', 'M6V', 'M7V', 'M8V'

In [93]:
if len(extract1)==len(extract2)==len(extract3):
    print('Yes')
else:
    print('No')

Yes


In [94]:
len(extract1)
len(extract2)
len(extract3)

180

180

180

In [41]:
post=[]
post_rows1=[]
for i in range(1,181):
    for y in data_rows[i]['Postal Code']:
        #print(y)
        post.append(y)
        post_rows=' '.join(post)
    post_rows1.append(post_rows)
print(post_rows1)

['M 1 A', 'M 1 A M 2 A', 'M 1 A M 2 A M 3 A', 'M 1 A M 2 A M 3 A M 4 A', 'M 1 A M 2 A M 3 A M 4 A M 5 A', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B M 2 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B M 2 B M 3 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B M 2 B M 3 B M 4 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B M 2 B M 3 B M 4 B M 5 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B M 2 B M 3 B M 4 B M 5 B M 6 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B M 2 B M 3 B M 4 B M 5 B M 6 B M 7 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A M 6 A M 7 A M 8 A M 9 A M 1 B M 2 B M 3 B M 4 B M 5 B M 6 B M 7 B M 8 B', 'M 1 A M 2 A M 3 A M 4 A M 5 A

In [42]:
post_rows1[1]

'M 1 A M 2 A'

In [43]:
data_rows[1]["Postal Code"]
for x in data_rows:
    print(x)

'M1A'

{}
{'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}
{'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}
{'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}
{'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}
{'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}
{'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}
{'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}
{'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}
{'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}
{'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}
{'Postal Code': 'M2B', 'Borough': 'Not assigned', 'Neighbourhood': 'Not ass

In [46]:
type(data_rows)
type(data_rows[1])
data_rows[1]['Postal Code']
data_rows[6]['Postal Code']

list

dict

'M1A'

'M6A'

In [47]:
for x in data1:
    print(x)

<td>Not assigned
</td>


In [48]:
data1.values()

dict_values([[{}, {'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}, {'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}, {'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}, {'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}, {'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}, {'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}, {'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}, {'Postal Code': 'M2B', 'Borough': 'Not assigned', 

In [49]:
#data1.keys()
#data1.items()
data1.get('Postal Code')

In [51]:
type(data_rows)
eg=data_rows[1]
eg
eg['Postal Code']

list

{'Postal Code': 'M1A',
 'Borough': 'Not assigned',
 'Neighbourhood': 'Not assigned'}

'M1A'

In [52]:
for i in range(1,181):
    for y in data_rows[i].values():
        print(y)

M1A
Not assigned
Not assigned
M2A
Not assigned
Not assigned
M3A
North York
Parkwoods
M4A
North York
Victoria Village
M5A
Downtown Toronto
Regent Park, Harbourfront
M6A
North York
Lawrence Manor, Lawrence Heights
M7A
Downtown Toronto
Queen's Park, Ontario Provincial Government
M8A
Not assigned
Not assigned
M9A
Etobicoke
Islington Avenue, Humber Valley Village
M1B
Scarborough
Malvern, Rouge
M2B
Not assigned
Not assigned
M3B
North York
Don Mills
M4B
East York
Parkview Hill, Woodbine Gardens
M5B
Downtown Toronto
Garden District, Ryerson
M6B
North York
Glencairn
M7B
Not assigned
Not assigned
M8B
Not assigned
Not assigned
M9B
Etobicoke
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
M1C
Scarborough
Rouge Hill, Port Union, Highland Creek
M2C
Not assigned
Not assigned
M3C
North York
Don Mills
M4C
East York
Woodbine Heights
M5C
Downtown Toronto
St. James Town
M6C
York
Humewood-Cedarvale
M7C
Not assigned
Not assigned
M8C
Not assigned
Not assigned
M9C
Etobicoke
Eringate, Bl

In [53]:
data1.items()

dict_items([(<td>Not assigned
</td>, [{}, {'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}, {'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}, {'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}, {'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}, {'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}, {'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}, {'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}, {'Postal Code': 'M2B', 'Bo

In [81]:
# printing original list 
print("The original list is : " + str(data_rows)) 

The original list is : [{}, {'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}, {'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}, {'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}, {'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}, {'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}, {'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}, {'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}, {'Postal Code': 'M2B', 'Borough': 'Not a

### The code made after testing

In [56]:
#table_req = data_soup.find("div", attrs={"class": "mw-parser-output"})
table_req = data_soup.find("div", attrs={"class": "mw-parser-output"})
#table_req
table_data = table_req.tbody.find_all("tr")  # contains many rows
# Get all the headings of Lists
headings = []
for td in table_data[0].find_all("th"):
    # remove any newlines and extra spaces from left and right
    headings.append(td.text.replace('\n', ' ').strip())
    #headings.append(td.b.text.replace('\n', ' ').strip())
print(headings)

['Postal Code', 'Borough', 'Neighbourhood']


In [57]:
table_data

[<tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighbourhood
 </th></tr>,
 <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>,
 <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>,
 <tr>
 <td>M5A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Regent Park, Harbourfront
 </td></tr>,
 <tr>
 <td>M6A
 </td>
 <td>North York
 </td>
 <td>Lawrence Manor, Lawrence Heights
 </td></tr>,
 <tr>
 <td>M7A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Queen's Park, Ontario Provincial Government
 </td></tr>,
 <tr>
 <td>M8A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>,
 <tr>
 <td>M9A
 </td>
 <td>Etobicoke
 </td>
 <td>Islington Avenue, Humber Valley Village
 </td></tr>,
 <tr>
 <td>M1B
 </td>
 <td>Scarborough
 </td>
 <td>Malvern, Rouge
 </td></tr>,
 <tr>
 <td>M2B
 </td>
 <td>Not assigned
 </td>

In [68]:
data1 = {}
data_rows = []
for tr in table_data: # find all tr's from table's tbody
    t_row = {}
    # Each table row is stored in the form of a dictionary
       # find all td's in tr and zip it with headings
    for td, th in zip(tr.find_all("td"), headings): 
        t_row[th] = td.text.replace('\n', '').strip()
    data_rows.append(t_row)
    # Put the data for the table with its heading.
data1[td] = data_rows
print(data1)

{<td>Not assigned
</td>: [{}, {'Postal Code': 'M1A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M2A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'}, {'Postal Code': 'M4A', 'Borough': 'North York', 'Neighbourhood': 'Victoria Village'}, {'Postal Code': 'M5A', 'Borough': 'Downtown Toronto', 'Neighbourhood': 'Regent Park, Harbourfront'}, {'Postal Code': 'M6A', 'Borough': 'North York', 'Neighbourhood': 'Lawrence Manor, Lawrence Heights'}, {'Postal Code': 'M7A', 'Borough': 'Downtown Toronto', 'Neighbourhood': "Queen's Park, Ontario Provincial Government"}, {'Postal Code': 'M8A', 'Borough': 'Not assigned', 'Neighbourhood': 'Not assigned'}, {'Postal Code': 'M9A', 'Borough': 'Etobicoke', 'Neighbourhood': 'Islington Avenue, Humber Valley Village'}, {'Postal Code': 'M1B', 'Borough': 'Scarborough', 'Neighbourhood': 'Malvern, Rouge'}, {'Postal Code': 'M2B', 'Borough': 'Not

In [59]:
type(data1)
type(data_rows)

dict

list

In [176]:
extract1=[d['Postal Code'] for d in data_rows if 'Postal Code' in d]
extract2=[d['Borough'] for d in data_rows if 'Borough' in d]
extract3=[d['Neighbourhood'] for d in data_rows if 'Neighbourhood' in d]

In [177]:
if len(extract1)==len(extract2)==len(extract3):
    print('Yes')
else:
    print('No')

Yes


In [178]:
toronto_table=pd.DataFrame(columns=['Postal Code','Borough','Neighbourhood'])

In [179]:
toronto_table

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [180]:
toronto_table['Postal Code']=extract1
toronto_table['Borough']=extract2
toronto_table['Neighbourhood']=extract3

In [181]:
toronto_table.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [182]:
toronto_table.shape

(180, 3)

In [183]:
toronto_table.describe()

Postal Code       Borough Neighbourhood
count          180           180           180
unique         180            11           100
top            M8X  Not assigned  Not assigned
freq             1            77            77

In [184]:
toronto_table.isnull().sum()

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

In [185]:
toronto_table['Borough'].replace(to_replace='Not assigned',value=np.nan,inplace=True)
toronto_table['Neighbourhood'].replace('Not assigned',np.nan,inplace=True)

In [186]:
toronto_table['Borough']
toronto_table['Neighbourhood']

0                   NaN
1                   NaN
2            North York
3            North York
4      Downtown Toronto
             ...       
175                 NaN
176                 NaN
177                 NaN
178           Etobicoke
179                 NaN
Name: Borough, Length: 180, dtype: object

0                                                    NaN
1                                                    NaN
2                                              Parkwoods
3                                       Victoria Village
4                              Regent Park, Harbourfront
                             ...                        
175                                                  NaN
176                                                  NaN
177                                                  NaN
178    Mimico NW, The Queensway West, South of Bloor,...
179                                                  NaN
Name: Neighbourhood, Length: 180, dtype: object

In [187]:
toronto_table['Borough'].isnull().sum()
toronto_table['Neighbourhood'].isnull().sum()

77

77

In [188]:
#toronto_table['Borough'].dropna(axis=0,inplace=True)
#toronto_table['Neighbourhood'].dropna(axis=0,inplace=True)
toronto_table.dropna(axis=0,inplace=True)

In [189]:
toronto_table['Borough'].isnull().sum()
toronto_table['Neighbourhood'].isnull().sum()
toronto_table.isnull().sum()

0

0

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

In [190]:
toronto_table.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [195]:
toronto_table.reset_index(drop=True)

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [196]:
toronto_table.shape  

(103, 3)

- #### After removing all the nan values from the dataframe there is a total of 103 rows and 3 columns with necessary data

In [227]:
toronto_table.to_csv('toronto_wikitable.csv',index=False)

In [208]:
!pip install geocoder

In [212]:
toronto_table['Postal Code']

2      M3A
3      M4A
4      M5A
5      M6A
6      M7A
      ... 
160    M8X
165    M4Y
168    M7Y
169    M8Y
178    M8Z
Name: Postal Code, Length: 103, dtype: object

In [216]:
os.chdir('D:\Data science\DS Coursera IBM\Capstone project\Toronto project')
os.getcwd()

'D:\\Data science\\DS Coursera IBM\\Capstone project\\Toronto project'

In [214]:
geocoder.google('M8Z, Toronto, Ontario')#.format(toronto_table['Postal Code']))

<[REQUEST_DENIED] Google - Geocode [empty]>

In [215]:
#import geocoder # import geocoder
# initialize your variable to None
#lat_lng_coords = None
# loop until you get the coordinates
#while(lat_lng_coords is None):
#    g = geocoder.google('{}, Toronto, Ontario'.format(toronto_table['Postal Code']))
#    lat_lng_coords = g.latlng
#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

In [218]:
latnlong=pd.read_csv('Geospatial_Coordinates.csv')
latnlong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [225]:
latnlong[(latnlong['Postal Code']=='M5A')] 

Postal Code  Latitude  Longitude
53         M5A  43.65426 -79.360636

- To check whether the data is correctly joined wrt to Postal code and it matches right for M5A postal code lat and log coordinates to the lat and long coordinates of toronto table
- So both the tables merged correctly 

In [221]:
latnlong.shape

(103, 3)

In [222]:
merged_left = pd.merge(left=toronto_table, right=latnlong, how='left', left_on='Postal Code', right_on='Postal Code')
merged_left

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [223]:
merged_left.isnull().sum()

Postal Code      0
Borough          0
Neighbourhood    0
Latitude         0
Longitude        0
dtype: int64

In [226]:
merged_left.shape

(103, 5)

In [230]:
merged_left.to_csv('toronto_with_latnlong.csv',index=False)